indra/media_plugins/cef/linux_volume_catcher_pw.cpp

```diff
@@ -0,0 +1,451 @@
+{
+	return gSymbolGrabber.grabSymbols({ pw_dso_name });
+}
+
+void VolumeCatcherImpl::init()
+{
+	// debugClear();
+	// debugPrint("init\n");
+	
+	mGotSyms = loadsyms("libpipewire-0.3.so.0");
+	if (!mGotSyms) return;
+
+	llpw_init(NULL, NULL);
+
+	mThreadLoop = llpw_thread_loop_new("SL Plugin Volume Adjuster", NULL);
+	if (!mThreadLoop) return;
+
+	pwLock();
+
+	mContext = llpw_context_new(
+		llpw_thread_loop_get_loop(mThreadLoop), NULL, 0
+	);
+	if (!mContext) return pwUnlock();
+
+	mCore = llpw_context_connect(mContext, NULL, 0);
+	if (!mCore) return pwUnlock();
+
+	mRegistry = pw_core_get_registry(mCore, PW_VERSION_REGISTRY, 0);
+
+	// debugPrint("got registry\n");
+
+	spa_zero(mRegistryListener);
+
+	pw_registry_add_listener(
+		mRegistry, &mRegistryListener, &REGISTRY_EVENTS, this
+	);
+
+	llpw_thread_loop_start(mThreadLoop);
+
+	pwUnlock();
+
+	// debugPrint("started thread loop\n");
+}
+
+void VolumeCatcherImpl::cleanup()
+{
+	mChildNodesMutex.lock();
+	for (auto* childNode : mChildNodes) {
+		childNode->destroy();
+	}
+	mChildNodes.clear();
+	mChildNodesMutex.unlock();
+
+	pwLock();
+	if (mRegistry) llpw_proxy_destroy((struct pw_proxy*)mRegistry);
+	spa_zero(mRegistryListener);
+	if (mCore) llpw_core_disconnect(mCore);
+	if (mContext) llpw_context_destroy(mContext);
+	pwUnlock();
+
+	if (!mThreadLoop) return;
+
+	llpw_thread_loop_stop(mThreadLoop);
+	llpw_thread_loop_destroy(mThreadLoop);
+
+	// debugPrint("cleanup done\n");
+}
+
+void VolumeCatcherImpl::pwLock() {
+	if (!mThreadLoop) return;
+	llpw_thread_loop_lock(mThreadLoop);
+}
+
+void VolumeCatcherImpl::pwUnlock() {
+	if (!mThreadLoop) return;
+	llpw_thread_loop_unlock(mThreadLoop);
+}
+
+// #define INV_LERP(a, b, v) (v - a) / (b - a)
+
+// #include <sys/time.h>
+
+void VolumeCatcherImpl::ChildNode::updateVolume()
+{
+	if (!mActive) return;
+
+	F32 volume = std::clamp(mImpl->mVolume, 0.0f, 1.0f);
+	// F32 pan = std::clamp(mImpl->mPan, -1.0f, 1.0f);
+
+	// debugClear();
+	// struct timeval time;
+	// gettimeofday(&time, NULL);
+	// double t = (double)time.tv_sec + (double)(time.tv_usec / 1000) / 1000;
+	// debugPrint("time is %f\n", t);
+	// F32 pan = std::sin(t * 2.0d);
+	// debugPrint("pan is %f\n", pan);
+
+	// uint32_t channels = 2;
+	// float volumes[channels];
+	// volumes[1] = INV_LERP(-1.0f, 0.0f, pan) * volume; // left
+	// volumes[0] = INV_LERP(1.0f, 0.0f, pan) * volume; // right
+
+	uint32_t channels = 1;
+	float volumes[channels];
+	volumes[0] = volume;
+
+	uint8_t buffer[512];
+
+	spa_pod_builder builder;
+	spa_pod_builder_init(&builder, buffer, sizeof(buffer));
+
+	spa_pod_frame frame;
+	spa_pod_builder_push_object(&builder, &frame, SPA_TYPE_OBJECT_Props, SPA_PARAM_Props);
+	spa_pod_builder_prop(&builder, SPA_PROP_channelVolumes, 0);
+	spa_pod_builder_array(&builder, sizeof(float), SPA_TYPE_Float, channels, volumes);
+	spa_pod* pod = static_cast<spa_pod*>(spa_pod_builder_pop(&builder, &frame));
+
+	mImpl->pwLock();
```

Comment:
It would be better to wrap those locks into some sort of RAII, just in case something rogue throws

Reply:
Definitely use RAII for locking and unlocking.